File: src/main/java/io/github/emcw/EMCMap.java
package io.github.emcw;

import io.github.emcw.caching.CacheOptions;
import io.github.emcw.caching.CacheStrategy;
import io.github.emcw.map.api.*;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.TimeUnit;

public class EMCMap {
    @Getter final String mapName;

    @Setter(AccessLevel.PRIVATE) public Towns Towns = null;
    @Setter(AccessLevel.PRIVATE) public Nations Nations = null;
    @Setter(AccessLevel.PRIVATE) public Players Players = null;
    @Setter(AccessLevel.PRIVATE) public Residents Residents = null;
    //@Setter(AccessLevel.PRIVATE) public GPS GPS = null;

    final CacheOptions lazyOpts = new CacheOptions(2, TimeUnit.SECONDS, CacheStrategy.LAZY);
    final CacheOptions timedOpts = new CacheOptions(3, TimeUnit.MINUTES, CacheStrategy.TIME_BASED);

    EMCMap(@NotNull KnownMap map) {
        this.mapName = map.getName();

        initCaches();
        prefill();

        //setGPS(new GPS(this));
    }

//    EMCMap(String mapName, CacheOptions mapDataCache, CacheOptions playerDataCache, boolean prefill) {
//        this.mapName = mapName;
//
//        setTowns(new Towns(mapDataCache));
//        setNations(new Nations(mapDataCache));
//        setResidents(new Residents(mapDataCache));
//        setPlayers(new Players(this, playerDataCache));
//
//        if (prefill) prefill();
//
//        //setGPS(new GPS(this));
//    }

    private void initCaches() {
        Towns = new Towns(timedOpts);
        Nations = new Nations(timedOpts);
        Residents = new Residents(timedOpts);
        Players = new Players(this, lazyOpts);
    }

    private void prefill() {
        Towns.forceUpdate();
        Nations.forceUpdate();
        Residents.forceUpdate();
        Players.forceUpdate();
    }
}

File: src/main/kotlin/io/github/emcw/EMCMap.kt
package io.github.emcw

import io.github.emcw.caching.CacheOptions
import io.github.emcw.caching.CacheStrategy
import io.github.emcw.map.api.Nations
import io.github.emcw.map.api.Players
import io.github.emcw.map.api.Residents
import io.github.emcw.map.api.Towns
import lombok.*
import java.util.concurrent.TimeUnit

class EMCMap {
    @Getter
    val mapName: String

    @Setter(AccessLevel.PRIVATE)
    var Towns: Towns? = null

    @JvmField
    @Setter(AccessLevel.PRIVATE)
    var Nations: Nations? = null

    @JvmField
    @Setter(AccessLevel.PRIVATE)
    var Players: Players? = null

    @Setter(AccessLevel.PRIVATE)
    var Residents: Residents? = null
    val lazyOpts = CacheOptions(2, TimeUnit.SECONDS, CacheStrategy.LAZY)
    val timedOpts = CacheOptions(3, TimeUnit.MINUTES, CacheStrategy.TIME_BASED)

    constructor(mapName: String) {
        this.mapName = mapName
        initCaches()
    }

    @JvmOverloads
    constructor(mapName: String, mapDataCache: CacheOptions, playerDataCache: CacheOptions, prefill: Boolean = false) {
        this.mapName = mapName
        setTowns(Towns(this, mapDataCache))
        setNations(Nations(this, mapDataCache))
        setResidents(Residents(this, mapDataCache))
        setPlayers(Players(this, playerDataCache))
        if (prefill) prefill()
    }

    private fun initCaches() {
        setTowns(Towns(this, timedOpts))
        setNations(Nations(this, timedOpts))
        setResidents(Residents(this, timedOpts))
        setPlayers(Players(this, lazyOpts))
        prefill()
    }

    private fun prefill() {
        Towns!!.forceUpdate()
        Nations!!.forceUpdate()
        Residents!!.forceUpdate()
        Players!!.forceUpdate()
    }
}